monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/MergePrioritizedListObservable.scala

```diff
@@ -0,0 +1,157 @@
+
+    val numberOfObservables = sources.size
+
+    val lock = new AnyRef
+    var isDone = false
+
+    // NOTE: We use arrays and other mutable structures here to be as performant as possible.
+
+    // MUST BE synchronized by `lock`
+    var lastAck = Continue: Future[Ack]
+
+    case class PQElem(data: A, promise: Promise[Ack], priority: Int) extends Comparable[PQElem] {
+      override def compareTo(o: PQElem): Int =
+        priority.compareTo(o.priority)
+    }
+
+    val pq = new PriorityBlockingQueue[PQElem](sources.size)
+
+    // MUST BE synchronized by `lock`
+    var completedCount = 0
+
+    // MUST BE synchronized by `lock`
+    def rawOnNext(a: A): Future[Ack] = {
+      if (isDone) Stop else out.onNext(a)
+    }
+
+    def processNext(): Future[Ack] = {
+      val e = pq.remove()
+      val fut = rawOnNext(e.data)
+      e.promise.completeWith(fut)
+      fut
+    }
+
+    // MUST BE synchronized by `lock`
+    def signalOnNext(): Future[Ack] =
+      lock.synchronized {
+        lastAck = lastAck match {
+          case Continue => processNext()
+          case Stop => Stop
+          case async =>
+            async.flatMap {
+              // async execution, we have to re-sync
+              case Continue => lock.synchronized(processNext())
+              case Stop => Stop
+            }
+        }
+
+        lastAck
+      }
+
+    def signalOnError(ex: Throwable): Unit =
+      lock.synchronized {
+        if (!isDone) {
+          isDone = true
+          out.onError(ex)
+          lastAck = Stop
+          completePromises()
+        }
+      }
+
+    def signalOnComplete(): Unit =
+      lock.synchronized {
+        completedCount += 1
+
+        if (completedCount == numberOfObservables && !isDone) {
+          lastAck match {
+            case Continue =>
+              isDone = true
+              out.onComplete()
+              completePromises()
+            case Stop =>
+              () // do nothing
+            case async =>
+              async.onComplete {
+                case Success(Continue) =>
+                  lock.synchronized {
+                    if (!isDone) {
+                      isDone = true
+                      out.onComplete()
+                      completePromises()
+                    }
+                  }
+                case _ =>
+                  () // do nothing
+              }
+          }
+
+          lastAck = Stop
+        }
+      }
+
+    def completePromises(): Unit = {
+      pq.iterator().asScala.foreach(e => e.promise.complete(Success(Stop)))
+    }
+
+    val composite = CompositeCancelable()
+
+    sources.zip(priorities).foreach { pair =>
+      composite += pair._1.unsafeSubscribeFn(new Subscriber[A] {
+        implicit val scheduler: Scheduler = out.scheduler
+
+        def onNext(elem: A): Future[Ack] = {
+          if (isDone) {
+            Stop
+          } else {
+            pq.add(PQElem(elem, Promise(), pair._2))
+            signalOnNext()
```

Comment:
is this returning the right thing? It looks like this is going to recieve the ack for the next value sent, regardless of whether it's `elem` or not. Should this return the `.future` side of the Promise created on line 143?

Reply:
Yes you're right, good catch, will fix.

Reply:
Fixed